### PR TITLE
Add Pulsar support

### PR DESCRIPTION
**Pull request by BewareMyPower**
This PR is to add support for [Apache Pulsar](http://pulsar.apache.org), which is a cloud-native, distributed messaging and streaming platform.

This PR introduced two classes:
- `PulsarIODataset`: Read streaming messages from Pulsar topics
- `PulsarWriter`: Write messages to Pulsar topics

Since Pulsar has a lot in common with Kafka, the `PulsarIODataset` implementation is similar to `KafkaGroupIODataset` except it cannot check the partition EOF.

Different from the simple `write_kafka` function, `PulsarWriter` has three ops: init, write messages asynchronously, flush pending messages.

### CODE REVIEW

**Comment by kvignesh1420** [comment]
@BewareMyPower thanks for the PR!

**Review by kvignesh1420 — COMMENTED** [review]
There seems to be a build issue:

```console

caused by: ['/v/tensorflow_io/core/python/ops/libtensorflow_io.so: cannot open shared object file: No such file or directory', 'bazel-bin/tensorflow_io/core/python/ops/libtensorflow_io.so: undefined symbol: BIO_f_base64']

```
@BewareMyPower can you please check the dependencies while building the CPP package?
Most of the times this might be due to `SSL` like dep's.

**Review by kvignesh1420 — COMMENTED** [review]
Also, I see that the build fails on windows:

```console
ERROR: C:/users/runneradmin/_bazel_runneradmin/m4q24vrm/external/pulsar/BUILD.bazel:18:1: C++ compilation of rule '@pulsar//:pulsar' failed (Exit 2): cl.exe failed: error executing command 
2020-11-19T10:56:03.1693092Z C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winbase.h(9461): warning C5105: macro expansion producing 'defined' has undefined behavior
2020-11-19T10:56:03.1695937Z C:\users\runneradmin\_bazel_runneradmin\m4q24vrm\execroot\org_tensorflow_io\bazel-out\x64_windows-fastbuild\bin\external\pulsar\pulsar-client-cpp\lib\PulsarApi.pb.h(11295): error C2059: syntax error: 'constant'
2020-11-19T10:56:03.1698133Z   cd C:/users/runneradmin/_bazel_runneradmin/m4q24vrm/execroot/org_tensorflow_io
2020-11-19T10:56:03.1700436Z C:\users\runneradmin\_bazel_runneradmin\m4q24vrm\execroot\org_tensorflow_io\bazel-out\x64_windows-fastbuild\bin\external\pulsar\pulsar-client-cpp\lib\PulsarApi.pb.h(11296): error C2238: unexpected token(s) preceding ';'
```

If the fix is not available/straight forward, we can take this up in the future PR's and remove the pulsar_op build for windows.

**Comment by BewareMyPower** [comment]
@kvignesh1420 Thanks for the review. I'll take some time to look into these issues.

The windows build failure looks like the ProtoBuf's issues since `PulsarApi.pb.h` is generated by ProtoBuf. I'm nearly new to Bazel and don't know how to add Protobuf 2.6 to the dependency. So I just use the existed Protobuf 3 to generate code and it works well in my local environment (MacOS).

Except these failures you mentioned, I also found something more. The Pulsar service startup stuck at

```bash
until curl http://localhost:8080/metrics > /dev/null 2>&1 ; do sleep 1; done
```

It may need a max loop count limit like what other tests do.

As for the failure of `Test 3.8 Linux`:

```
sed -i '' 's/brokerDeleteInactiveTopicsFrequencySeconds=.*/brokerDeleteInactiveTopicsFrequencySeconds=86400/' conf/standalone.conf
767
sed: can't read s/brokerDeleteInactiveTopicsFrequencySeconds=.*/brokerDeleteInactiveTopicsFrequencySeconds=86400/: No such file or directory
```

It's the incompatibility issue of `sed` command. I just want to change a default config of the config file. In MacOS, `sed 's/xxx/yyy/' file` doesn't work but `sed '' 's/xxx/yyy/ file` works. But in Linux environment it may be the contrast.

Could you give some help on how to solve this problem?

**Comment by BewareMyPower** [comment]
> There seems to be a build issue:
> 
> ```
> caused by: ['/v/tensorflow_io/core/python/ops/libtensorflow_io.so: cannot open shared object file: No such file or directory', 'bazel-bin/tensorflow_io/core/python/ops/libtensorflow_io.so: undefined symbol: BIO_f_base64']
> ```
> 
> @BewareMyPower can you please check the dependencies while building the CPP package?
> Most of the times this might be due to `SSL` like dep's.

Yeah, the pulsar-client-cpp has ssl and crypto dependencies. Just forgot to add these to the `deps`.

**Review by terrytangyuan — COMMENTED** [review]
Thanks for your contribution!

**Comment by yongtang** [comment]
@BewareMyPower You can replace `sed -i '' xxx` with `sed -i.bak xxx` and that will work for both Linux and macOS I think.

**Comment by kvignesh1420** [comment]
@BewareMyPower sure. 

Also, as @yongtang said, you can use:
 ```
sed -i.bak 's/brokerDeleteInactiveTopicsFrequencySeconds=.*/brokerDeleteInactiveTopicsFrequencySeconds=86400/' conf/standalone.conf
```
for replacement. I tried it on CentOS 7 & macOS and it works fine. This will also create a `conf/standalone.conf.bak` file as a backup.

**Comment by yongtang** [comment]
@BewareMyPower The most likely scenario for errors like the following on windows:
```
2020-11-19T10:56:03.1695937Z C:\users\runneradmin\_bazel_runneradmin\m4q24vrm\execroot\org_tensorflow_io\bazel-out\x64_windows-fastbuild\bin\external\pulsar\pulsar-client-cpp\lib\PulsarApi.pb.h(11295): error C2059: syntax error: 'constant'
```

is caused by a symbol defined in `Windows.h`. The last similar error I saw was on pubsub where `OPTIONAL` has already been defined in Windows header files. As such we have to use the following tricks in 
https://github.com/tensorflow/io/blob/c314881702a996c8e802c30fba958edaacba4b9e/tensorflow_io/core/kernels/pubsub_kernels.cc#L20-L25

For the error in pulsar I tend to believe we can inject the following to PulsarApi.pb.h (or before inclusion of PulsarApi.pb.h) as well if we can find the symbol conflict:
```
 #if defined(_MSC_VER) 
 #include <Windows.h> 
 #undef XXXCONFLICTSYMBOLXXX
 #endif 
```

**Comment by BewareMyPower** [comment]
Thanks @yongtang and @kvignesh1420 for the great help!

As for the Windows issue, I just remembered that Pulsar service only support running on Linux and MacOS. A Pulsar service is started in `tests/pulsar_test.sh` so it will always fail on Windows.

Is there any way to skip the pulsar_ops when running Windows test CI?

**Comment by yongtang** [comment]
@BewareMyPower Can you make the following changes and give it a try?

1. In WORKSPACE, replace `pulsar` with the following and add `dlfn-win32`:
```
http_archive(
    name = "dlfcn-win32",
    build_file = "//third_party:dlfcn-win32.BUILD",
    sha256 = "f18a412e84d8b701e61a78252411fe8c72587f52417c1ef21ca93604de1b9c55",
    strip_prefix = "dlfcn-win32-1.2.0",
    urls = [
        "https://storage.googleapis.com/mirror.tensorflow.org/github.com/dlfcn-win32/dlfcn-win32/archive/v1.2.0.tar.gz",
        "https://github.com/dlfcn-win32/dlfcn-win32/archive/v1.2.0.tar.gz",
    ],
)

http_archive(
    name = "pulsar",
    build_file = "//third_party:pulsar.BUILD",
    patch_cmds = [
        "cp pulsar-common/src/main/proto/PulsarApi.proto pulsar-client-cpp/lib",
        "sed -i.bak 's/define PULSAR_DEFINES_H_/define PULSAR_DEFINES_H_\\'$'\\n''#if defined(_MSC_VER)\\'$'\\n''#include <Windows.h>\\'$'\\n''#undef ERROR\\'$'\\n''#endif/g' pulsar-client-cpp/include/pulsar/defines.h",
        "sed -i.bak 's/define LIB_ACKGROUPINGTRACKER_H_/define LIB_ACKGROUPINGTRACKER_H_\\'$'\\n''#include <pulsar\\/defines.h>/g' pulsar-client-cpp/lib/AckGroupingTracker.h",
    ],
    sha256 = "08f19ca6d6353751ff0661403b16b71425bf7ada3d8835a38e426ae303b0e385",
    strip_prefix = "pulsar-2.6.1",
    urls = [
        "https://github.com/apache/pulsar/archive/v2.6.1.tar.gz",
    ],
)
```

2. Add third_party/dlfcn-win32.BUILD:
```
# Description:
#   dlfcn-win32

package(default_visibility = ["//visibility:public"])

licenses(["notice"])  # MIT

exports_files(["COPYING"])

cc_library(
    name = "dlfcn-win32",
    srcs = [
        "dlfcn.c",
        "dlfcn.h",
    ],
    hdrs = [],
    defines = [],
    includes = [
        ".",
    ],
    deps = [],
)
```

3. In third_party/pulsar.BUILD apply the following patch (add `"WIN32_LEAN_AND_MEAN"` and  `"PULSAR_STATIC"` to `defines`, and conditionally depends on `dlfcn-win32`)
```diff
diff --git a/third_party/pulsar.BUILD b/third_party/pulsar.BUILD
index bc3258b..2cc7d58 100644
--- a/third_party/pulsar.BUILD
+++ b/third_party/pulsar.BUILD
@@ -35,7 +35,11 @@ cc_library(
         "-msse4.2",
         "-mpclmul",
     ],
-    defines = ["_PULSAR_VERSION_=\\\"2.6.1\\\""],
+    defines = [
+        "_PULSAR_VERSION_=\\\"2.6.1\\\"",
+        "WIN32_LEAN_AND_MEAN",
+        "PULSAR_STATIC",
+    ],
     includes = [
         "pulsar-client-cpp",
         "pulsar-client-cpp/include",
@@ -47,5 +51,10 @@ cc_library(
         "@boost",
         "@curl",
         "@zstd",
-    ],
+    ] + select({
+        "@bazel_tools//src/conditions:windows": [
+            "@dlfcn-win32",
+        ],
+        "//conditions:default": [],
+    }),
 )

```

I think with above changes Windows build issue will be resolved.

File: tests/test_pulsar/pulsar_test.sh [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+# ==============================================================================
+
+set -e
+set -o pipefail
+
+VERSION=2.6.1
+TAR_FILE="apache-pulsar-${VERSION}-bin.tar.gz"
+
+echo "Downloading pulsar ${VERSION}"
+if [[ ! -f apache-pulsar-2.6.1-bin.tar.gz ]]; then
```
Comment (terrytangyuan) [inline tests/test_pulsar/pulsar_test.sh:24]: Reuse VERSION here and below

Comment (BewareMyPower) [reply tests/test_pulsar/pulsar_test.sh]: Ok, I've fixed it.

**Comment by BewareMyPower** [comment]
Thanks for @yongtang 's detailed instruction. I've applied it now. The tests may still fail because the issue that Pulsar service cannot run on Windows. But let's see how many CI checks would succeed first.

**Comment by BewareMyPower** [comment]
It looks like the Pulsar service cannot start in 30 seconds with CI environment, which makes tests fail. I'll take a look at this issue.

**Comment by BewareMyPower** [comment]
The Ubuntu build failure is still caused by `undefined symbol: BIO_f_base64`. It seems that the added dependencies didn't work:

```
        "@boringssl//:crypto",
        "@boringssl//:ssl",
```

The windows build failed by link error:

> 2020-11-20T16:07:29.0860149Z   C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.28.29333/bin/HostX64/x64/cl.exe /nologo /DCOMPILER_MSVC /DNOMINMAX /D_WIN32_WINNT=0x0601 /D_CRT_SECURE_NO_DEPRECATE /D_CRT_SECURE_NO_WARNINGS /bigobj /Zm500 /EHsc /wd4351 /wd4291 /wd4250 /wd4996 /I. /Ibazel-out/x64_windows-fastbuild/bin /Iexternal/local_config_tf /Ibazel-out/x64_windows-fastbuild/bin/external/local_config_tf /Iexternal/com_github_googleapis_google_cloud_cpp /Ibazel-out/x64_windows-fastbuild/bin/external/com_github_googleapis_google_cloud_cpp /Iexternal/com_google_absl /Ibazel-out/x64_windows-fastbuild/bin/external/com_google_absl /Iexternal/boringssl /Ibazel-out/x64_windows-fastbuild/bin/external/boringssl /Iexternal/curl /Ibazel-out/x64_windows-fastbuild/bin/external/curl /Iexternal/zlib /Ibazel-out/x64_windows-fastbuild/bin/external/zlib /Iexternal/com_github_google_crc32c /Ibazel-out/x64_windows-fastbuild/bin/external/com_github_google_crc32c /Iexternal/nlohmann_json_lib /Ibazel-out/x64_windows-fastbuild/bin/external/nlohmann_json_lib /Iexternal/local_config_tf/include /Ibazel-out/x64_windows-fastbuild/bin/external/local_config_tf/include /Iexternal/boringssl/src/include /Ibazel-out/x64_windows-fastbuild/bin/external/boringssl/src/include /Iexternal/curl/include /Ibazel-out/x64_windows-fastbuild/bin/external/curl/include /Iexternal/curl/lib /Ibazel-out/x64_windows-fastbuild/bin/external/curl/lib /Iexternal/zlib /Ibazel-out/x64_windows-fastbuild/bin/external/zlib /Iexternal/zlib/contrib/minizip /Ibazel-out/x64_windows-fastbuild/bin/external/zlib/contrib/minizip /Iexternal/com_github_google_crc32c/include /Ibazel-out/x64_windows-fastbuild/bin/external/com_github_google_crc32c/include /Iexternal/nlohmann_json_lib/include /Ibazel-out/x64_windows-fastbuild/bin/external/nlohmann_json_lib/include /DCURL_STATICLIB /DBUILDING_LIBCURL /DHAVE_CONFIG_H /DCURL_DISABLE_FTP /DCURL_DISABLE_NTLM /DHAVE_LIBZ /DHAVE_ZLIB_H /DCURL_MAX_WRITE_SIZE=16384 /D__CLANG_SUPPORT_DYN_ANNOTATION__ /showIncludes /MD /Od /Z7 /wd4117 -D__DATE__="redacted" -D__TIMESTAMP__="redacted" -D__TIME__="redacted" -D_GLIBCXX_USE_CXX11_ABI=0 /Zc:preprocessor -std=c++14 /DEIGEN_STRONG_INLINE=inline -DTENSORFLOW_MONOLITHIC_BUILD /DPLATFORM_WINDOWS /DEIGEN_HAS_C99_MATH /DTENSORFLOW_USE_EIGEN_THREADPOOL /DEIGEN_AVOID_STL_ARRAY /Iexternal/gemmlowp /wd4018 /wd4577 /DNOGDI /UTF_COMPILE_LIBRARY /DNDEBUG /Fobazel-out/x64_windows-fastbuild/bin/tensorflow_io/core/plugins/gs/_objs/gs/gcs_filesystem.obj /c tensorflow_io/core/plugins/gs/gcs_filesystem.cc
> 2020-11-20T16:07:34.9369975Z ERROR: D:/a/io/io/tensorflow_io/core/BUILD:682:1: Linking of rule '//tensorflow_io/core:python/ops/libtensorflow_io.so' failed (Exit 1120): link.exe failed: error executing command

**Comment by yongtang** [comment]
@BewareMyPower Can you apply the following patch and give it a try (not it is applied to boringssl's repo entry):
```diff
diff --git a/WORKSPACE b/WORKSPACE
index 45e81d6..9995e69 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -405,6 +405,9 @@ http_archive(
 
 http_archive(
     name = "boringssl",
+    patch_cmds = [
+        """sed -i.bak 's/bio.c",/bio.c","src\\/decrepit\\/bio\\/base64_bio.c",/g' BUILD.generated.bzl""",
+    ],
     sha256 = "1188e29000013ed6517168600fc35a010d58c5d321846d6a6dfee74e4c788b45",
     strip_prefix = "boringssl-7f634429a04abc48e2eb041c81c5235816c96514",
     urls = [
```

**Comment by BewareMyPower** [comment]
Thanks @yongtang for the help. It also makes me more familiar with Bazel :)

**Comment by BewareMyPower** [comment]
The tests failed because Pulsar and Kafka both use ZooKeeper for metadata management.

```
16:25:34.531 [main] ERROR org.apache.pulsar.zookeeper.LocalBookkeeperEnsemble - Exception while instantiating ZooKeeper
java.net.BindException: Address already in use
16:25:34.536 [main] ERROR org.apache.pulsar.PulsarStandaloneStarter - Failed to start pulsar service.
java.io.IOException: java.net.BindException: Address already in use
```

The CI logs have no similar logs because Pulsar doesn't flush logs to disk immediately by default.

I'll change the Pulsar's ZK settings and try again.

**Review by yongtang — APPROVED** [review]
LGTM

**Comment by yongtang** [comment]
@BewareMyPower The PR LGTM. Thanks for the contribution!

Also, we normally have tutorials published on tensorflow.org's official documentation distempered (e.g., https://www.tensorflow.org/io/tutorials/kafka). Given Pulsar is a new services, would you like to contribute a tutorial in a follow up PR?

**Comment by BewareMyPower** [comment]
@yongtang Okay. I'll create a tutorial later.

File: tests/test_pulsar_eager.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
```
Comment (kvignesh1420) [inline tests/test_pulsar_eager.py:1]: Please change the year to 2020

File: tests/test_pulsar/pulsar_test.sh [hunk 3]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env bash
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
```
Comment (kvignesh1420) [inline tests/test_pulsar/pulsar_test.sh:2]: Please change the year to 2020

File: tensorflow_io/core/python/experimental/pulsar_writer_ops.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
```
Comment (kvignesh1420) [inline tensorflow_io/core/python/experimental/pulsar_writer_ops.py:1]: Please change the year to 2020

File: tensorflow_io/core/python/experimental/pulsar_dataset_ops.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
```
Comment (kvignesh1420) [inline tensorflow_io/core/python/experimental/pulsar_dataset_ops.py:1]: Please change the year to 2020

File: tensorflow_io/core/ops/pulsar_ops.cc [hunk 6]
```diff
@@ -0,0 +1,71 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```
Comment (kvignesh1420) [inline tensorflow_io/core/ops/pulsar_ops.cc:1]: Please change the year to 2020

File: tensorflow_io/core/kernels/pulsar_kernel.cc [hunk 7]
```diff
@@ -0,0 +1,355 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```
Comment (kvignesh1420) [inline tensorflow_io/core/kernels/pulsar_kernel.cc:1]: Please change the year to 2020

**Review by kvignesh1420 — COMMENTED** [review]
@BewareMyPower,  please check the comments for minor changes. Otherwise, it looks good to me. Thanks!

**Review by kvignesh1420 — APPROVED** [review]
LGTM. Thanks for the contribution @BewareMyPower

**Pull request merged**